File: java/src/wildling/TemplateJson.java
package wildling;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Minimal JSON parser for wildling template files (stdlib only).
 * Supports objects, arrays, strings, numbers, booleans, and null.
 */
final class TemplateJson {
    private final String text;
    private int pos;

    private TemplateJson(String text) {
        this.text = text;
        this.pos = 0;
    }

    static Object parse(String text) {
        TemplateJson parser = new TemplateJson(text);
        Object value = parser.parseValue();
        parser.skipWhitespace();
        if (parser.pos != parser.text.length()) {
            throw new IllegalArgumentException("Unexpected trailing JSON content");
        }
        return value;
    }

    @SuppressWarnings("unchecked")
    static Map<String, Object> parseObject(String text) {
        Object value = parse(text);
        if (!(value instanceof Map)) {
            throw new IllegalArgumentException("Template root must be a JSON object");
        }
        return (Map<String, Object>) value;
    }

    private Object parseValue() {
        skipWhitespace();
        if (pos >= text.length()) {
            throw new IllegalArgumentException("Unexpected end of JSON");
        }
        char c = text.charAt(pos);
        if (c == '{') {
            return parseObjectValue();
        }
        if (c == '[') {
            return parseArray();
        }
        if (c == '"') {
            return parseString();
        }
        if (c == 't' || c == 'f') {
            return parseBoolean();
        }
        if (c == 'n') {
            return parseNull();
        }
        if (c == '-' || Character.isDigit(c)) {
            return parseNumber();
        }
        throw new IllegalArgumentException("Unexpected character at " + pos);
    }

    private Map<String, Object> parseObjectValue() {
        expect('{');
        Map<String, Object> object = new LinkedHashMap<>();
        skipWhitespace();
        if (peek('}')) {
            pos++;
            return object;
        }
        while (true) {
            skipWhitespace();
            String key = parseString();
            skipWhitespace();
            expect(':');
            Object value = parseValue();
            object.put(key, value);
            skipWhitespace();
            if (peek('}')) {
                pos++;
                return object;
            }
            expect(',');
        }
    }

    private List<Object> parseArray() {
        expect('[');
        List<Object> array = new ArrayList<>();
        skipWhitespace();
        if (peek(']')) {
            pos++;
            return array;
        }
        while (true) {
            array.add(parseValue());
            skipWhitespace();
            if (peek(']')) {
                pos++;
                return array;
            }
            expect(',');
        }
    }

    private String parseString() {
        expect('"');
        StringBuilder out = new StringBuilder();
        while (pos < text.length()) {
            char c = text.charAt(pos++);
            if (c == '"') {
                return out.toString();
            }
            if (c == '\\') {
                if (pos >= text.length()) {
                    throw new IllegalArgumentException("Unterminated escape");
                }
                char esc = text.charAt(pos++);
                switch (esc) {
                    case '"':
                    case '\\':
                    case '/':
                        out.append(esc);
                        break;
                    case 'b':
                        out.append('\b');
                        break;
                    case 'f':
                        out.append('\f');
                        break;
                    case 'n':
                        out.append('\n');
                        break;
                    case 'r':
                        out.append('\r');
                        break;
                    case 't':
                        out.append('\t');
                        break;
                    case 'u':
                        if (pos + 4 > text.length()) {
                            throw new IllegalArgumentException("Invalid unicode escape");
                        }
                        int code = Integer.parseInt(text.substring(pos, pos + 4), 16);
                        out.append((char) code);
                        pos += 4;
                        break;
                    default:
                        throw new IllegalArgumentException("Invalid escape \\" + esc);
                }
            } else {
                out.append(c);
            }
        }
        throw new IllegalArgumentException("Unterminated string");
    }

    private Object parseNumber() {
        int start = pos;
        if (peek('-')) {
            pos++;
        }
        while (pos < text.length() && Character.isDigit(text.charAt(pos))) {
            pos++;
        }
        boolean isDouble = false;
        if (peek('.')) {
            isDouble = true;
            pos++;
            while (pos < text.length() && Character.isDigit(text.charAt(pos))) {
                pos++;
            }
        }
        if (pos < text.length() && (text.charAt(pos) == 'e' || text.charAt(pos) == 'E')) {
            isDouble = true;
            pos++;
            if (peek('+') || peek('-')) {
                pos++;
            }
            while (pos < text.length() && Character.isDigit(text.charAt(pos))) {
                pos++;
            }
        }
        String raw = text.substring(start, pos);
        if (isDouble) {
            return Double.parseDouble(raw);
        }
        return Long.parseLong(raw);
    }

    private Boolean parseBoolean() {
        if (text.startsWith("true", pos)) {
            pos += 4;
            return Boolean.TRUE;
        }
        if (text.startsWith("false", pos)) {
            pos += 5;
            return Boolean.FALSE;
        }
        throw new IllegalArgumentException("Invalid boolean at " + pos);
    }

    private Object parseNull() {
        if (text.startsWith("null", pos)) {
            pos += 4;
            return null;
        }
        throw new IllegalArgumentException("Invalid null at " + pos);
    }

    private void skipWhitespace() {
        while (pos < text.length()) {
            char c = text.charAt(pos);
            if (c == ' ' || c == '\n' || c == '\r' || c == '\t') {
                pos++;
            } else {
                break;
            }
        }
    }

    private boolean peek(char expected) {
        return pos < text.length() && text.charAt(pos) == expected;
    }

    private void expect(char expected) {
        skipWhitespace();
        if (!peek(expected)) {
            throw new IllegalArgumentException("Expected '" + expected + "' at " + pos);
        }
        pos++;
    }
}

File: kotlin/src/wildling/TemplateJson.kt
package wildling

/**
 * Minimal JSON parser for wildling template files (stdlib only).
 * Supports objects, arrays, strings, numbers, booleans, and null.
 */
internal class TemplateJson private constructor(private val text: String) {
    private var pos: Int = 0

    companion object {
        fun parse(text: String): Any? {
            val parser = TemplateJson(text)
            val value = parser.parseValue()
            parser.skipWhitespace()
            if (parser.pos != parser.text.length) {
                throw IllegalArgumentException("Unexpected trailing JSON content")
            }
            return value
        }

        @Suppress("UNCHECKED_CAST")
        fun parseObject(text: String): Map<String, Any?> {
            val value = parse(text)
            if (value !is Map<*, *>) {
                throw IllegalArgumentException("Template root must be a JSON object")
            }
            return value as Map<String, Any?>
        }
    }

    private fun parseValue(): Any? {
        skipWhitespace()
        if (pos >= text.length) {
            throw IllegalArgumentException("Unexpected end of JSON")
        }
        val c = text[pos]
        return when {
            c == '{' -> parseObjectValue()
            c == '[' -> parseArray()
            c == '"' -> parseString()
            c == 't' || c == 'f' -> parseBoolean()
            c == 'n' -> parseNull()
            c == '-' || c in '0'..'9' -> parseNumber()
            else -> throw IllegalArgumentException("Unexpected character at $pos")
        }
    }

    private fun parseObjectValue(): Map<String, Any?> {
        expect('{')
        val obj = linkedMapOf<String, Any?>()
        skipWhitespace()
        if (peek('}')) {
            pos++
            return obj
        }
        while (true) {
            skipWhitespace()
            val key = parseString()
            skipWhitespace()
            expect(':')
            obj[key] = parseValue()
            skipWhitespace()
            if (peek('}')) {
                pos++
                return obj
            }
            expect(',')
        }
    }

    private fun parseArray(): List<Any?> {
        expect('[')
        val array = mutableListOf<Any?>()
        skipWhitespace()
        if (peek(']')) {
            pos++
            return array
        }
        while (true) {
            array.add(parseValue())
            skipWhitespace()
            if (peek(']')) {
                pos++
                return array
            }
            expect(',')
        }
    }

    private fun parseString(): String {
        expect('"')
        val out = StringBuilder()
        while (pos < text.length) {
            val c = text[pos++]
            when {
                c == '"' -> return out.toString()
                c == '\\' -> {
                    if (pos >= text.length) {
                        throw IllegalArgumentException("Unterminated escape")
                    }
                    when (val esc = text[pos++]) {
                        '"', '\\', '/' -> out.append(esc)
                        'b' -> out.append('\b')
                        'f' -> out.append('\u000c')
                        'n' -> out.append('\n')
                        'r' -> out.append('\r')
                        't' -> out.append('\t')
                        'u' -> {
                            if (pos + 4 > text.length) {
                                throw IllegalArgumentException("Invalid unicode escape")
                            }
                            val code = text.substring(pos, pos + 4).toInt(16)
                            out.append(code.toChar())
                            pos += 4
                        }
                        else -> throw IllegalArgumentException("Invalid escape \\$esc")
                    }
                }
                else -> out.append(c)
            }
        }
        throw IllegalArgumentException("Unterminated string")
    }

    private fun parseNumber(): Number {
        val start = pos
        if (peek('-')) {
            pos++
        }
        while (pos < text.length && text[pos].isDigit()) {
            pos++
        }
        var isDouble = false
        if (peek('.')) {
            isDouble = true
            pos++
            while (pos < text.length && text[pos].isDigit()) {
                pos++
            }
        }
        if (pos < text.length && (text[pos] == 'e' || text[pos] == 'E')) {
            isDouble = true
            pos++
            if (peek('+') || peek('-')) {
                pos++
            }
            while (pos < text.length && text[pos].isDigit()) {
                pos++
            }
        }
        val raw = text.substring(start, pos)
        return if (isDouble) raw.toDouble() else raw.toLong()
    }

    private fun parseBoolean(): Boolean {
        when {
            text.startsWith("true", pos) -> {
                pos += 4
                return true
            }
            text.startsWith("false", pos) -> {
                pos += 5
                return false
            }
            else -> throw IllegalArgumentException("Invalid boolean at $pos")
        }
    }

    private fun parseNull(): Any? {
        if (text.startsWith("null", pos)) {
            pos += 4
            return null
        }
        throw IllegalArgumentException("Invalid null at $pos")
    }

    private fun skipWhitespace() {
        while (pos < text.length) {
            when (text[pos]) {
                ' ', '\n', '\r', '\t' -> pos++
                else -> return
            }
        }
    }

    private fun peek(expected: Char): Boolean =
        pos < text.length && text[pos] == expected

    private fun expect(expected: Char) {
        skipWhitespace()
        if (!peek(expected)) {
            throw IllegalArgumentException("Expected '$expected' at $pos")
        }
        pos++
    }
}
